python/cudf/cudf/tests/test_copying.py

```diff
@@ -52,3 +53,324 @@ def test_null_copy():
     col = Series(np.arange(2049))
     col[:] = None
     assert len(col) == 2049
+
+
+# TODO: Make use of set_option context manager
+# once https://github.com/rapidsai/cudf/issues/12736
+# is resolved.
+
+
+def test_series_setitem_cow_on():
+    original_cow_setting = cudf.get_option("copy_on_write")
+    cudf.set_option("copy_on_write", True)
+    actual = cudf.Series([1, 2, 3, 4, 5])
+    new_copy = actual.copy(deep=False)
+
+    actual[1] = 100
+    assert_eq(actual, cudf.Series([1, 100, 3, 4, 5]))
+    assert_eq(new_copy, cudf.Series([1, 2, 3, 4, 5]))
+
+    actual = cudf.Series([1, 2, 3, 4, 5])
+    new_copy = actual.copy(deep=False)
+
+    actual[slice(0, 2, 1)] = 100
+    assert_eq(actual, cudf.Series([100, 100, 3, 4, 5]))
+    assert_eq(new_copy, cudf.Series([1, 2, 3, 4, 5]))
+
+    new_copy[slice(2, 4, 1)] = 300
+    assert_eq(actual, cudf.Series([100, 100, 3, 4, 5]))
+    assert_eq(new_copy, cudf.Series([1, 2, 300, 300, 5]))
+
+    actual = cudf.Series([1, 2, 3, 4, 5])
+    new_copy = actual.copy(deep=False)
+
+    new_copy[slice(2, 4, 1)] = 300
+    assert_eq(actual, cudf.Series([1, 2, 3, 4, 5]))
+    assert_eq(new_copy, cudf.Series([1, 2, 300, 300, 5]))
+
+    new_slice = actual[2:]
+    assert new_slice._column.base_data._ptr == actual._column.base_data._ptr
+    new_slice[0:2] = 10
+    assert_eq(new_slice, cudf.Series([10, 10, 5], index=[2, 3, 4]))
+    assert_eq(actual, cudf.Series([1, 2, 3, 4, 5]))
+    cudf.set_option("copy_on_write", original_cow_setting)
+
+
+def test_series_setitem_cow_off():
+    original_cow_setting = cudf.get_option("copy_on_write")
+    cudf.set_option("copy_on_write", False)
+    actual = cudf.Series([1, 2, 3, 4, 5])
+    new_copy = actual.copy(deep=False)
+
+    actual[1] = 100
+    assert_eq(actual, cudf.Series([1, 100, 3, 4, 5]))
+    assert_eq(new_copy, cudf.Series([1, 100, 3, 4, 5]))
+
+    actual = cudf.Series([1, 2, 3, 4, 5])
+    new_copy = actual.copy(deep=False)
+
+    actual[slice(0, 2, 1)] = 100
+    assert_eq(actual, cudf.Series([100, 100, 3, 4, 5]))
+    assert_eq(new_copy, cudf.Series([100, 100, 3, 4, 5]))
+
+    new_copy[slice(2, 4, 1)] = 300
+    assert_eq(actual, cudf.Series([100, 100, 300, 300, 5]))
+    assert_eq(new_copy, cudf.Series([100, 100, 300, 300, 5]))
+
+    actual = cudf.Series([1, 2, 3, 4, 5])
+    new_copy = actual.copy(deep=False)
+
+    new_copy[slice(2, 4, 1)] = 300
+    assert_eq(actual, cudf.Series([1, 2, 300, 300, 5]))
+    assert_eq(new_copy, cudf.Series([1, 2, 300, 300, 5]))
+
+    new_slice = actual[2:]
+    assert new_slice._column.base_data._ptr == actual._column.base_data._ptr
+    new_slice[0:2] = 10
+    assert_eq(new_slice, cudf.Series([10, 10, 5], index=[2, 3, 4]))
+    assert_eq(actual, cudf.Series([1, 2, 10, 10, 5]))
+    cudf.set_option("copy_on_write", original_cow_setting)
+
+
+def test_multiple_series_cow():
```

Comment:
This test is... complicated. But I get why it's complicated, and I guess it is worthwhile to have at least one such integration-level test. Maybe an extra comment explaining how at each step of modification one more series is being uncoupled, and then at the end deletion needs to not affect still extant series?

Reply:
Done, added detailed comments.